File: node/src/util/io.mjs
import { mkdir, readFile, writeFile, rename, lstat, unlink, open } from "node:fs/promises";
import { dirname, resolve, sep } from "node:path";
import { existsSync } from "node:fs";
import { spawnSync } from "node:child_process";
import { randomBytes } from "node:crypto";
import { homedir } from "node:os";

export async function readSafe(path) {
  try { return await readFile(path, "utf8"); }
  catch (e) { if (e.code === "ENOENT") return ""; throw e; }
}

function isPermErr(e) { return e?.code === "EACCES" || e?.code === "EPERM"; }

async function isSymlink(p) {
  try { return (await lstat(p)).isSymbolicLink(); }
  catch { return false; }
}

function underHome(p, home) {
  const r = resolve(p);
  const h = resolve(home);
  return r === h || r.startsWith(h + sep);
}

function shellQuote(s) { return `'${String(s).replace(/'/g, `'\\''`)}'`; }

async function reclaim(path, home, err) {
  if (process.platform === "win32" || typeof process.getuid !== "function") return false;
  if (!existsSync(path)) return false;
  if (await isSymlink(path)) {
    err.write(`pmsec: refusing to chown symlink ${path}; remove or replace it manually.\n`);
    return false;
  }
  if (!underHome(path, home)) {
    err.write(`pmsec: refusing to chown ${path} outside HOME (${home}); fix ownership manually.\n`);
    return false;
  }
  err.write(`pmsec: ${path} not writable; running \`sudo chown -h $(id -u):$(id -g) ${shellQuote(path)}\`. You may be prompted for your password.\n`);
  const r = spawnSync("sudo", ["chown", "-h", `${process.getuid()}:${process.getgid()}`, path], { stdio: "inherit" });
  return r.status === 0;
}

async function atomicReplace(path, text) {
  const tmp = `${path}.${process.pid}.${randomBytes(4).toString("hex")}.tmp`;
  let fh;
  try {
    fh = await open(tmp, "wx", 0o600);
    await fh.writeFile(text, "utf8");
    try { await fh.sync(); } catch {}
    await fh.close();
    fh = null;
    await rename(tmp, path);
  } catch (e) {
    if (fh) { try { await fh.close(); } catch {} }
    try { await unlink(tmp); } catch {}
    throw e;
  }
}

export async function writeAtomic(path, text, { backup = true, home = homedir(), err = process.stderr } = {}) {
  await mkdir(dirname(path), { recursive: true });
  if (existsSync(path) && await isSymlink(path)) {
    const e = new Error(`refusing to write through symlink ${path}`);
    e.code = "ELOOP";
    throw e;
  }
  if (backup && existsSync(path) && !existsSync(path + ".bak")) {
    try { await atomicReplace(path + ".bak", await readFile(path, "utf8")); }
    catch (e) {
      const target = existsSync(path + ".bak") ? path + ".bak" : path;
      if (!isPermErr(e) || !(await reclaim(target, home, err))) throw e;
      await atomicReplace(path + ".bak", await readFile(path, "utf8"));
    }
  }
  try { await atomicReplace(path, text); }
  catch (e) {
    if (!isPermErr(e) || !(await reclaim(path, home, err))) throw e;
    await atomicReplace(path, text);
  }
}

File: python/src/pmsec/util/io.py
from __future__ import annotations

import os
import secrets
import shlex
import stat
import subprocess
import sys
from pathlib import Path


def _is_perm_err(exc: BaseException) -> bool:
    return isinstance(exc, PermissionError)


def _is_symlink(path: Path) -> bool:
    try:
        return stat.S_ISLNK(path.lstat().st_mode)
    except FileNotFoundError:
        return False


def _under_home(path: Path, home: Path) -> bool:
    try:
        target = path.resolve(strict=False)
        root = home.resolve(strict=False)
    except OSError:
        return False
    try:
        target.relative_to(root)
        return True
    except ValueError:
        return False


def _reclaim(path: Path, home: Path, err) -> bool:
    if sys.platform == "win32" or not hasattr(os, "geteuid"):
        return False
    if not path.exists():
        return False
    if _is_symlink(path):
        err.write(f"pmsec: refusing to chown symlink {path}; remove or replace it manually.\n")
        return False
    if not _under_home(path, home):
        err.write(f"pmsec: refusing to chown {path} outside HOME ({home}); fix ownership manually.\n")
        return False
    quoted = shlex.quote(str(path))
    err.write(
        f"pmsec: {path} not writable; running `sudo chown -h $(id -u):$(id -g) {quoted}`. "
        "You may be prompted for your password.\n"
    )
    err.flush()
    r = subprocess.run(["sudo", "chown", "-h", f"{os.geteuid()}:{os.getegid()}", str(path)])
    return r.returncode == 0


def _atomic_replace(path: Path, text: str) -> None:
    parent = path.parent
    tmp = parent / f".{path.name}.{os.getpid()}.{secrets.token_hex(4)}.tmp"
    try:
        fd = os.open(tmp, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
        try:
            with os.fdopen(fd, "w", encoding="utf-8") as f:
                f.write(text)
                f.flush()
                try:
                    os.fsync(f.fileno())
                except OSError:
                    pass
        except BaseException:
            try:
                tmp.unlink()
            except FileNotFoundError:
                pass
            raise
        os.replace(tmp, path)
    except BaseException:
        try:
            tmp.unlink()
        except FileNotFoundError:
            pass
        raise


def write_atomic(path: Path, text: str, *, backup: bool = True, home: Path | None = None, err=sys.stderr) -> None:
    home = Path.home() if home is None else home
    path.parent.mkdir(parents=True, exist_ok=True)
    if path.exists() and _is_symlink(path):
        raise OSError(f"refusing to write through symlink {path}")
    if backup and path.exists():
        bak = path.with_suffix(path.suffix + ".bak")
        if not bak.exists():
            try:
                _atomic_replace(bak, path.read_text("utf-8"))
            except OSError as exc:
                if not _is_perm_err(exc) or not _reclaim(bak if bak.exists() else path, home, err):
                    raise
                _atomic_replace(bak, path.read_text("utf-8"))
    try:
        _atomic_replace(path, text)
    except OSError as exc:
        if not _is_perm_err(exc) or not _reclaim(path, home, err):
            raise
        _atomic_replace(path, text)
